secio/src/codec/secure_stream.rs

```diff
@@ -303,15 +316,51 @@ where
             self.nonce.drain(..n);
             frame.split_to(n);
         }
-        Ok(())
+        Ok(frame)
+    }
+
+    fn decode(&mut self) -> Poll<(), SecioError> {
+        if self.current_decode.is_none() {
+            return Ok(Async::Ready(()));
+        }
+        let data = self.current_decode.clone().unwrap();
+        let t = match tokio_threadpool::blocking(|| self.decode_inner(data.clone())) {
```

Comment:
data is cloned twice? maybe we can take the data then set it back if fail to decode?

Reply:
https://github.com/tokio-rs/tokio/issues/1172